aws/retry/adaptive.go

```diff
@@ -0,0 +1,173 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/ratelimit"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+)
+
+const (
+	// DefaultRequestCost is the cost of a single request from the adaptive
+	// rate limited token bucket.
+	DefaultRequestCost uint = 1
+)
+
+// DefaultThrottles provides the set of errors considered throttle errors that
+// are checked by default.
+var DefaultThrottles = []IsErrorThrottle{
+	ThrottleErrorCode{
+		Codes: DefaultThrottleErrorCodes,
+	},
+}
+
+// AdaptiveModeOptions provides the functional options for configuring the
+// adaptive retry mode, and delay behavior.
+type AdaptiveModeOptions struct {
+	FailOnNoAttemptTokens bool
```

Comment:
Is this option called out in the specification, should we just rely on context cancellation behavior instead?

Reply:
This option is explicitly called out as a SDK **should** provide option to immediately fail if no attempt rate limited tokens are available. This kind of behavior is different than Context cancel, since context is generally time based at a higher level. Whereas this option is action based, and will have no impact until the client is begin throttled, and the token bucket has been drained.

I think the context cancel is a good pattern for the operation TTL overall, but work as well to limit to per attempt dynamic behavior.